File: sorting_in_linear_time/counting_sort.py
import random


def sort(data):
    count_table = [0 for _ in range(max(data) + 1)]

    for d in data:
        count_table[d] += 1

    print("--- counting table before accumlating ---")
    print(count_table)

    for i in range(len(count_table) - 1):
        count_table[i + 1] += count_table[i]

    print("--- counting table after accumlating ---")
    print(count_table)

    auxiliary_table = [-1 for _ in range(len(data))]
    for i in reversed(range(len(data))):
        target_data = data[i]
        idx_of_target = count_table[target_data] - 1
        auxiliary_table[idx_of_target] = target_data
        count_table[target_data] -= 1

    data.clear()
    data.extend(auxiliary_table)


if __name__ == '__main__':
    data = [random.randint(1, 20) for _ in range(40)]
    
    print("--- before ---")
    print(data)

    print("--- sort ---")
    sort(data)

    print("--- after ---")
    print(data)

File: sorting_in_linear_time/counting_sort.go
//go:build ignore

package main

import (
	"fmt"
	"math/rand"
)

func createData() []int {
	data := make([]int, 40)

	for i := 0; i < 40; i++ {
		data[i] = rand.Intn(20)
	}

	return data
}

func countingSort(data []int) {
	n := len(data)
	max := 0
	for i := 0; i < n; i++ {
		if data[i] > max {
			max = data[i]
		}
	}

	countingTable := make([]int, max+1)
	for i := 0; i < n; i++ {
		countingTable[data[i]] += 1
	}
	for i := 1; i < len(countingTable); i++ {
		countingTable[i] += countingTable[i-1]
	}
	fmt.Println("--- counting table ---")
	fmt.Println(countingTable)

	temporaryData := make([]int, len(data))
	for i := n - 1; i >= 0; i-- {
		targetData := data[i]
		idxToSet := countingTable[targetData] - 1
		temporaryData[idxToSet] = targetData
		countingTable[targetData] -= 1
	}
	fmt.Println("--- temporary data ---")
	fmt.Println(temporaryData)

	data = data[:0]
	data = append(data, temporaryData...)
}

func main() {
	data := createData()

	fmt.Println("--- before ---")
	fmt.Println(data)

	fmt.Println("--- sort ---")
	countingSort(data)

	fmt.Println("--- after ---")
	fmt.Println(data)
}
